include/deal.II/base/mutable_bind.h

```diff
@@ -0,0 +1,298 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_base_mutable_bind_h
+#define dealii_base_mutable_bind_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/patterns.h>
+#include <deal.II/base/std_cxx14/utility.h>
+#include <deal.II/base/std_cxx17/tuple.h>
+
+#include <tuple>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Utilities
+{
+  /**
+   * A mutable version of std::bind, that binds all arguments of a function
+   * pointer to a stored tuple, and allows you to update the tuple between
+   * calls.
+   *
+   * An example usage of this class is through the helper function
+   * mutable_bind() that creates a MutableBind object on the fly, based on its
+   * arguments:
+   *
+   * @code
+   * void my_function(const int &a, const double &b);
+   *
+   * auto bound = mutable_bind(my_function, 1, 2.0);
+   *
+   * bound(); // will execute my_function(1, 2.0);
+   *
+   * bound.set_parameters(2, 3.0);
+   * bound(); // will execute my_function(2, 3.0);
+   *
+   * bound.parse_parameters("3: 4.0");
+   * bound(); // will execute my_function(3, 4.0);
+   * @endcode
+   *
+   * @authors Luca Heltai, Matthias Maier, 2019.
+   */
+  template <typename ReturnType, class... FunctionArgs>
+  struct MutableBind
+  {
+    /**
+     * A typedef to the stored std::tuple type.
+     */
+    using TupleType = std::tuple<typename std::remove_cv<
+      typename std::remove_reference<FunctionArgs>::type>::type...>;
```

Comment:
Sorry for the potentially stupid question: Does the `std::remove_reference` mean that objects are always transferred as copies into `arguments` of the bound function? If so, then I believe that this not work for objects with no copy constructor (e.g. `Triangulation`)? Maybe this should be documented.

Otherwise... Cool class :-)

Reply:
Yes, that's precisely what it means. I'll add a note in the documentation. :)